overlord/devicestate/remodel.go

```diff
@@ -0,0 +1,349 @@
+  creates a store that uses that and refers to the new brand store
+
+* different brand/model, maybe different brand store => reregRemodel
+  similar to storeSwitchRemodel case after a first phase that performs
+  re-registration where the context plays registrationContext's role
+  (NOT IMPLEMENTED YET)
+
+*/
+
+// RemodelKind designates a kind of remodeling.
+type RemodelKind int
+
+const (
+	// same brand/model, brand store
+	UpdateRemodel RemodelKind = iota
+	// same brand/model, different brand store
+	StoreSwitchRemodel
+	// different brand/model, maybe different brand store
+	ReregRemodel
+)
+
+func (k RemodelKind) String() string {
+	switch k {
+	case UpdateRemodel:
+		return "revision update remodel"
+	case StoreSwitchRemodel:
+		return "store switch remodel"
+	case ReregRemodel:
+		return "re-registration remodel"
+	}
+	panic(fmt.Sprintf("internal error: unknown remodel kind: %d", k))
+}
+
+// ClassifyRemodel returns what kind of remodeling is going from oldModel to newModel.
+func ClassifyRemodel(oldModel, newModel *asserts.Model) RemodelKind {
+	if oldModel.BrandID() != newModel.BrandID() {
+		return ReregRemodel
+	}
+	if oldModel.Model() != newModel.Model() {
+		return ReregRemodel
+	}
+	if oldModel.Store() != newModel.Store() {
+		return StoreSwitchRemodel
+	}
+	return UpdateRemodel
+}
+
+type remodelCtxKey struct {
+	chgID string
+}
+
+func cachedRemodelCtx(chg *state.Change) (remodelContext, bool) {
+	key := remodelCtxKey{chg.ID()}
+	remodCtx, ok := chg.State().Cached(key).(remodelContext)
+	return remodCtx, ok
+}
+
+func cleanupRemodelCtx(chg *state.Change) {
+	chg.State().Cache(remodelCtxKey{chg.ID()}, nil)
+}
+
+// A remodelContext mediates the correct and isolated device state
+// access and evolution during a remodel.
+// All remodelContexts are at least a DeviceContext.
+type remodelContext interface {
+	Init(chg *state.Change) error
+	Finish() error
+	snapstate.DeviceContext
+
+	// initialDevice takes the current/initial device state
+	// when setting up the remodel context
+	initialDevice(device *auth.DeviceState)
+	// associate associates the remodel context with the change
+	// and caches it
+	associate(chg *state.Change)
+}
+
+// remodelCtx returns a remodeling context for the given transition.
+// It constructs and caches a dedicated store as needed as well.
+func remodelCtx(st *state.State, oldModel, newModel *asserts.Model) (remodelContext, error) {
+	var remodCtx remodelContext
+
+	devMgr := deviceMgr(st)
+
+	switch kind := ClassifyRemodel(oldModel, newModel); kind {
+	case UpdateRemodel:
+		// simple context for the simple case
+		remodCtx = &updateRemodelContext{baseRemodelContext{newModel}}
+	case StoreSwitchRemodel:
+		storeSwitchCtx := &newStoreRemodelContext{
+			baseRemodelContext: baseRemodelContext{newModel},
+			st:                 st,
+			deviceMgr:          devMgr,
+		}
+		storeSwitchCtx.store = devMgr.newStore(storeSwitchCtx.deviceBackend())
+		remodCtx = storeSwitchCtx
+	default:
```

Comment:
Perhaps add a comment here that Rereg falls into default right now (or `ReregRemodel: fallthrough // TODO`)